python_transport/wirepas_gateway/protocol/mqtt_wrapper.py

```diff
@@ -0,0 +1,217 @@
+                    self._client.publish(topic, payload, qos=qos, retain=retain)
+
+                    # FIX: read internal sockpairR as it is written but
+                    # never read as we don't use the internal paho loop
+                    # but we have spurious timeout / broken pipe from
+                    # this socket pair
+                    try:
+                        self._client._sockpairR.recv(1)
+                    except Exception:
+                        # This socket is not used at all, so if something is wrong,
+                        # not a big issue. Just keep going
+                        pass
+
+            except TimeoutError:
+                self.logger.debug("Timeout to send payload: {}".format(payload))
+                # In theory, mqtt client shouldn't loose the last packet
+                # If it is not the case, following line could be uncommented
+                # self._publish_queue.put((topic, payload, qos, retain))
+                raise
+            except queue.Empty:
+                # No more packet to publish
+                pass
+
+        if sock in r:
+            self._client.loop_read()
+
+        if sock in w:
+            self._client.loop_write()
+
+        self._client.loop_misc()
+
+    def _get_socket(self):
+        sock = self._client.socket()
+        if sock is not None:
+            return sock
+
+        self.logger.error("MQTT, unexpected disconnection")
+
+        # Socket is not opened anymore, try to reconnect
+        while True:
+            try:
+                self._client.reconnect()
+                break
+            except Exception:
+                # Retry to connect in 1 sec
+                sleep(1)
+
+        # Wait for socket to reopen
+        # Do it in polling as we are managing the thread ourself
+        # so no callback possible
+        while self._client.socket() is None:
+            sleep(1)
+
+        # Set options to new reopened socket
+        self._client.socket().setsockopt(socket.SOL_SOCKET, socket.SO_SNDBUF, 2048)
+        return self._client.socket()
+
+    def set_last_will(self, topic, data):
+        # Set Last wil message
+        self._client.will_set(topic, data, qos=2, retain=True)
+
+    def run(self):
+        while True:
+            try:
+                # Get client socket to select on it
+                # This function manage the reconnect
+                sock = self._get_socket()
+
+                self._do_select(sock)
+            except TimeoutError:
+                self.logger.error("Timeout in connection, force a reconnect")
+                self._client.reconnect()
+            except Exception:
+                # If an exception is not catched before this point
+                # All the transport module must be stopped in order to be fully
+                # restarted by the managing agent
+                self.logger.exception("Unexpected exception in MQTT wrapper Thread")
```

Comment:
I would prefer to see the exception breaking and on the outside of the loop you could add the cleanup code.

Reply:
Done